File: 001-500/0128. Longest Consecutive Sequence.java
// Medium
// Set
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/longest-consecutive-sequence/

import java.util.*;
class Solution {
    // Use a Set to remove duplicates and enable fast lookup
    // Iterate through the set, and for each number, check if it can be start 
    // If it can start, continue incrementing and checking for n+1, counting the sequence length
    // Update the max length found for each sequence
    public int longestConsecutive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int num : nums) {
            set.add(num);
        }
        int res = 0;
        for (int x : set) {
            // check start: set cannot contain the n - 1
            if (!set.contains(x - 1)) {
                int curNum = x;
                int curRes = 1;
                while (set.contains(curNum + 1)) {
                    curNum++;
                    curRes++;
                }
                res = Math.max(res, curRes);
            }
        }
        return res;
    }
}

File: 0001-0500/0128. Longest Consecutive Sequence.py
# Medium
# Greedy
# Time:O(n), Space:O(1)
# https://leetcode.cn/problems/longest-consecutive-sequence/

# Use set to do linear search, but we need to make sure the start is correct
# for each "potential" start num, no num-1

from typing import *

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        num_set = set(nums) # O(n)
        max_len = 0

        for num in num_set:
            if num - 1 not in num_set:
                cur = num
                cur_len = 1

                while cur + 1 in num_set:
                    cur += 1
                    cur_len += 1
                
                max_len = max(max_len, cur_len)
        
        return max_len
